README
Flrig is a transceiver control program for Amateur Radio use.  It does
not use any 3rd party transceiver control libraries.  It is a c++ pro-
gram that encapsulates each transceiver in it's own class.  Where ever
possible the transceiver class(s) use polymorphism to reuse code that
is portable across a series of transceivers.

The latest version can always be found at:

  http://www.w1hkj.com/

Join the linuxham, NBEMSham or win-fldigi Yahoo group for support, news 
and updates on W1HKJ software:

  http://groups.yahoo.com/group/linuxham/join
  http://groups.yahoo.com/group/NBEMSham/join
  http://groups.yahoo.com/group/win-fldigi/join



ChangeLog
2023-05-15  dave-w1hkj  <w1hkj@bellsouth.net>

	2365f15e: Version 2.0.0   * Update release
	18c37870: xml command button
	badde39a: button colors
	ad188cff: string vectors
	22e7d265: FT890 update
	1c209bb4: X6100 update
	b401cf13: trusdx seg fault
	d67bde7b: combo highlight
	6d65eec9: TS590S/SG
	71609348: Yaesu FT710
	2ed6af50: Elecraft BW interface
	22553a44: xmlrpc server
	91cde455: IC7700 set BW
	23644911: Kenwood meter readings
	8968415b: FTdx10 bandwidths
	dcacfc1f: Version 1.4.8

2023-03-10  Harry McGavran  <w5pny@w5pny.com>

	e0d968c6: Remove K4 debug lines

2023-03-08  dave-w1hkj  <w1hkj@bellsouth.net>

	fe2bbc5c: Window decoration
	1c597f1e: PTT via xmlrpc
	823d2cea: FC LEAVE event
	e257978b: socket address
	5897199f: cwiox
	08f45bef: thread priority
	b4fd41b1: Iconify

2023-02-04  Leif Sawyer  <ak.hepcat+gio@gmail.com>

	c26976a8: Rescale dialogs
	b4c30d05: fltk 1.4 update

2023-02-02  dave-w1hkj  <w1hkj@bellsouth.net>

	7cf542ae: cwio timing
	d30e60ca: Warnings

2023-01-29  Black Michael  <mdblack98@yahoo.com>

	315c62eb: FreqControl/fltk-1.4

2023-01-29  dave-w1hkj  <w1hkj@bellsouth.net>

	9a0e49d9: Morse table
	17611bfe: xmlrpc cwio ptt
	c59a4fcb: QDX Version 1.06
	484b65c7: K3 BW command
	2ec006aa: GCC 13
	973d73c3: SunSDR2
	5069393e: Icom 705
	0548b9f8: TS570 smeter
	5fb18d7b: Flex 1500 U/L filter
	8e72233e: Yaesu vfo
	ee9aa3e1: Get/Set preamp/att

2023-01-05  w5pny  <w5pny@w5pny.com>

	be02d3d2: Window Decoration Icon

2023-01-03  dave-w1hkj  <w1hkj@bellsouth.net>

	926830cc: Button logic
	11eb15d9: SunSDR2 PBT

2023-01-02  Fred-N8YX  <N8YX@neo.rr.com>

	cc107890: PCR1000 dsp

2022-12-30  dave-w1hkj  <w1hkj@bellsouth.net>

	39dd4434: Combo visibility

2022-12-29  Harry McGavran  <w5pny@w5pny.com>

	8e6a1e2c: serial guardlock

2022-12-18  dave-w1hkj  <w1hkj@bellsouth.net>

	f22ed3ff: xmlrpc get vfo

2022-12-16  Fred_N8YX  <N8YX@neo.rr.com>

	d36f7c3e: Add IC751 ICR71

2022-12-16  dave-w1hkj  <w1hkj@bellsouth.net>

	ed251ec3: PBT
	45c7a8d7: Meters
	2b0ec8fb: Bandwidth update
	93932da1: PBT xmlrpc interface
	3716765c: QDX adj
	fd5980f9: Deactivate Meters
	a3f2c8b5: Power SDR update
	f3776cf0: UI - Frequency Control behavior
	b4dc46fb: Icom pbt controls
	342702f2: FTdx1200 write delay

2022-11-25  Dave G0GBX  <g8kbvdave@googlemail.com>

	b7e89dda: TS870S

2022-11-24  Black Michael  <mdblack98@yahoo.com>

	1a0d1c6e: MinGW warnings
	497dc95b: Multiple monitor

2022-11-23  dave-w1hkj  <w1hkj@bellsouth.net>

	7f5b2a0c: Xmlrpc seg fault

2022-11-23  Black Michael  <mdblack98@yahoo.com>

	15b15dde: Meters dialog

2022-11-22  dave-w1hkj  <w1hkj@bellsouth.net>

	785f8a5a: trusdx
	d6e6e87f: RHEL 7 build
	4b618c05: Xiegu rigs
	36426959: FTdx10 FA/FB
	f8af12e8: FTdx101D power display

2022-11-19  Dave Tillman  <blennie96@verizon.net>

	7b4f9d9f: Fixes for IC-7610 Extra Controls - Band Tab
	29d6d303: Address compiler warning messages
	91a0affa: Remove FC focus throws from Elecraft rig UIs
	90f71237: ULL Data Type Changes

2022-11-12  dave-w1hkj  <w1hkj@bellsouth.net>

	b5ce7bac: FTdx10 meters
	436c2038: QDX
	85bd2d8a: FT857
	9d9300f8: powersdr
	5fffbdd6: icom save  upper/lower

2022-10-30  Dave Tillman  <blennie96@verizon.net>

	98c6ca2d: User Interface - Frequency Control
	8ca1ba0b: ICOM IC-7610 PBT Slider fixes

2022-10-26  dave-w1hkj  <w1hkj@bellsouth.net>

	335221b6: Slider Widget Focus
	dcd105f2: Lower/Upper display
	bcff9f6d: Icom swap flag
	139b0659: SunSDR Pro/DX
	6697533e: BW UI Updates
	9ad2763e: serial-mods
	9daa203c: KENWOOD
	4d169452: byte interval
	1e45615d: PowerSDR bug
	2fa13a65: pcr1000 mod
	5b3a875c: message icon
	73f0e34a: Extras
	e8c71a37: Trace macros
	343b3c78: FTdx10 autostart
	da22b23c: combo

2022-09-12  Dave Tillman  <blennie96@verizon.net>

	87288b21: User Interface - Combobox fixes

2022-09-09  dave-w1hkj  <w1hkj@bellsouth.net>

	8c12a060: (tr)uSDX
	a5668525: ftdx10 notch
	c7d97c4a: seg fault fix
	1a2321a8: Smeter/Freq poll
	c5704912: FTdx10
	11c76b7e: IC7851

2022-08-15  w5pny  <w5pny@w5pny.com>

	13c4be93: Add K4 transceiver

2022-08-06  dave-w1hkj  <w1hkj@bellsouth.net>

	f0056a55: TS 590 S/SG AM data mode
	f7e797e0: Version 1.4.7
	b1387f6b: Missing images
	1c28e513: Memory manager
	eb41b42d: Version 1.4.6
	93a25beb: sdr2-win
	f84403b6: IC9700 satellite support
	6422130b: Yaesu microphone gain
	5af2552b: Icom mode_type
	dd0c523e: Yaesu split
	bba23c7d: SunSDR2 CW bandwidths

2022-07-01  Andrew Hart  <andylhart@gmail.com>

	b6c69cdc: IC7600 mods

2022-05-29  dave-w1hkj  <w1hkj@bellsouth.net>

	f841d124: 7300 FM CW
	b5555ab1: CW keyer enhancements
	0abb7fc8: fsk fixes
	d19ce45e: QCX+
	3a999275: SunSDR2 tci bandwidth
	f7fd5260: FTdx10 Power out
	f14dba5b: Vd / Idd metering
	09dae3d7: FT101DxMP voltage scale
	c281d657: X6100 / G90
	65064a3b: SunSDR2
	c0a410fc: Serial trace

2022-04-10  Joe Attili (AA2IL)  <joeattili@gmail.com>

	e1321823: FT991A/FTdx3000 mods

2022-04-06  dave-w1hkj  <w1hkj@bellsouth.net>

	f4bd6b94: Compression
	b5e81e4f: TS570 Split
	fb320437: pts ports
	7a4f1123: FT450/D
	e0f35f78: FT817 defaults
	e299604d: Version 1.4.5

2022-03-25  Black Michael  <mdblack98@yahoo.com>

	745a55f9: PowerSDR

2022-03-23  dave-w1hkj  <w1hkj@bellsouth.net>

	db77a6c1: cwio xcvr compensation
	63ab1da3: FT817 startup
	52a1e5a3: trace mod
	8d0a23c2: Serial Trace
	b6ce40b9: TS990
	c8b4fbc8: IC F8101
	f7bb8b2c: testmode
	15d8025d: visible tab
	4d7e2f87: Variable PTT
	cbd1758d: TT566
	5744c28f: FT817mod
	a059301c: RX polling
	351a5a5a: FT 857/897/D
	8442cb1d: TABS tracking
	05ff6c50: freq control
	109ddb49: TT Eagle
	53183eb5: vfo AB
	1df769ad: Xiegu X6100
	104a9d6f: traced
	ff699399: Source structure
	f78f4b95: Start up values
	2b00f166: setthex
	baee9877: fskio idle
	e69c257e: Power SDR
	1957addc: namespace std
	d5a45aca: AGC controls
	029edeb1: IC705 update
	3e2ead62: FT 817
	d209e151: 7300 power table
	916234ab: CW keydown
	b85adab7: cwio fskio
	3bb8fde3: 10 W Power scale
	c968af13: FT1000MP-MKV
	1e75dcfa: G90 update
	757d9e3c: xmlrpc mod vol

src/rigs.cxx
// ----------------------------------------------------------------------------
// Copyright (C) 2014
//              David Freese, W1HKJ
//
// This file is part of flrig.
//
// flrig is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// flrig is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// ----------------------------------------------------------------------------

//=============================================================================
// Reference Transceiver unique drivers

#include "rigs.h"
#include "util.h"
#include "debug.h"
#include "rig_io.h"

rigbase			rig_null;

RIG_FT100D		rig_FT100D;
RIG_FT450		rig_FT450;
RIG_FT747		rig_FT747;
RIG_FT767		rig_FT767;
RIG_FT817		rig_FT817;
RIG_FT847		rig_FT847;
RIG_FT857D		rig_FT857D;
RIG_FT890		rig_FT890;
RIG_FT897D		rig_FT897D;
RIG_FT920		rig_FT920;
RIG_FT950		rig_FT950;
RIG_FT990		rig_FT990;
RIG_FT1000MP	rig_FT1000MP;
RIG_FT2000		rig_FT2000;
RIG_FTdx1200	rig_FTdx1200;
RIG_FTdx3000	rig_FTdx3000;
RIG_FT5000		rig_FT5000;

RIG_IC703		rig_IC703;
RIG_IC706MKIIG	rig_IC706MKIIG;
RIG_IC718		rig_IC718;
RIG_IC728		rig_IC728;
RIG_IC735		rig_IC735;
RIG_IC746		rig_IC746;
RIG_IC746PRO	rig_IC746PRO;
RIG_IC756		rig_IC756;
RIG_IC756PRO	rig_IC756PRO;
RIG_IC756PRO2	rig_IC756PRO2;
RIG_IC756PRO3	rig_IC756PRO3;
RIG_IC7000		rig_IC7000;
RIG_IC7100		rig_IC7100;
RIG_IC7200		rig_IC7200;
RIG_IC7600		rig_IC7600;
RIG_IC7700		rig_IC7700;
RIG_IC9100		rig_IC9100;
RIG_IC910H		rig_IC910H;

RIG_K2			rig_K2;
RIG_K3			rig_K3;

RIG_PCR1000		rig_PCR1000;
RIG_RAY152		rig_RAY152;

RIG_TS140		rig_TS140;
RIG_TS450S		rig_TS450S;
RIG_TS480HX		rig_TS480HX;
RIG_TS480SAT	rig_TS480SAT;
RIG_TS570		rig_TS570;
RIG_TS590S		rig_TS590S;
RIG_TS870S		rig_TS870S;
RIG_TS990		rig_TS990;
RIG_TS2000		rig_TS2000;

RIG_TT516		rig_TT516;
RIG_TT535		rig_TT535;
RIG_TT538		rig_TT538;
RIG_TT550		rig_TT550;
RIG_TT563		rig_TT563;
RIG_TT566		rig_TT566;
RIG_TT588		rig_TT588;
RIG_TT599		rig_TT599;

rigbase *rigs[] = {
	&rig_null,		// 0
	&rig_FT100D,	// 1
	&rig_FT450,		// 2
	&rig_FT747,		// 3
	&rig_FT767,		// 4
	&rig_FT817,		// 5
	&rig_FT847,		// 6
	&rig_FT857D,	// 7
	&rig_FT890,		// 8
	&rig_FT897D,	// 9
	&rig_FT920,		// 10
	&rig_FT950,		// 11
	&rig_FT990,		// 12
	&rig_FT1000MP,	// 13
	&rig_FT2000,	// 14
	&rig_FTdx1200,	// 15
	&rig_FTdx3000,	// 16
	&rig_FT5000,	// 17
	&rig_IC703,		// 18
	&rig_IC706MKIIG,// 19
	&rig_IC718,		// 20
	&rig_IC728,		// 21
	&rig_IC735,		// 22
	&rig_IC746,		// 23
	&rig_IC746PRO,	// 24
	&rig_IC756,		// 25
	&rig_IC756PRO,	// 26
	&rig_IC756PRO2,	// 27
	&rig_IC756PRO3,	// 28
	&rig_IC7000,	// 29
	&rig_IC7100,	// 30
	&rig_IC7200,	// 31
	&rig_IC7600,	// 32
	&rig_IC7700,	// 33
	&rig_IC9100,	// 34
	&rig_IC910H,	// 35
	&rig_K2,		// 36
	&rig_K3,		// 37
	&rig_PCR1000,	// 38
	&rig_RAY152,	// 39
	&rig_TS140,		// 40
	&rig_TS450S,	// 41
	&rig_TS480HX,	// 42
	&rig_TS480SAT,	// 43
	&rig_TS570,		// 44
	&rig_TS590S,	// 45
	&rig_TS870S,	// 46 wbx
	&rig_TS990,		// 47
	&rig_TS2000,	// 48
	&rig_TT516,		// 49
	&rig_TT535,		// 50
	&rig_TT538,		// 51
	&rig_TT550,		// 52
	&rig_TT563,		// 53
	&rig_TT566,		// 54
	&rig_TT588,		// 55
	&rig_TT599,		// 56
	NULL
};

//=============================================================================
